src/main/java/org/swisspush/reststorage/s3/S3FileSystemStorage.java

```diff
@@ -0,0 +1,386 @@
+                log.debug("Successfully opened '{}' which is {} bytes in size.", path, d.length);
+
+                final FileReadStream<io.vertx.core.buffer.Buffer> readStream = new FileReadStream<>(vertx, d.length, path, inputStream);
+                d.readStream = readStream;
+
+                final Runnable cleanUp = () -> {
+                    if (!readStream.isClosed()){
+                    readStream.close();
+                    }
+                    try {
+                        inputStream.close();
+                    } catch (IOException e) {
+                        log.debug("Failed to close input stream", e);
+                    }
+                };
+                d.closeHandler = v -> {
+                    log.debug("Resource got closed. Close file now '{}'", path);
+                    cleanUp.run();
+                };
+                d.addErrorHandler(event -> {
+                    log.error("Get resource failed.", exceptionFactory.newException("Close file now '" + path + "'", event));
+                    cleanUp.run();
+                });
+            } catch (IOException e) {
+                log.warn("Failed to open '{}' for read", path, e);
+                d.error = true;
+                d.errorMessage = e.getMessage();
+            }
+            handler.handle(d);
+            return;
+        }
+
+        if (Files.isDirectory(fullDirPath)) {
+            log.debug("Delegate directory listing of '{}'", path);
+            fileSystemDirLister.handleListingRequest(fullDirPath, offset, count, handler);
+        } else {
+            log.debug("No such file or directory '{}' ({})", path, fullDirPath);
+            Resource r = new Resource();
+            r.exists = false;
+            handler.handle(r);
+        }
+    }
+
+    @Override
+    public void put(String path, String etag, boolean merge, long expire, final Handler<Resource> handler) {
+        put(path, etag, merge, expire, "", LockMode.SILENT, 0, handler);
+    }
+
+    @Override
+    public void put(String path, String etag, boolean merge, long expire, String lockOwner, LockMode lockMode, long lockExpire, Handler<Resource> handler) {
+        final Path fullPath = canonicalize(path, false);
+        if (Files.exists(fullPath, LinkOption.NOFOLLOW_LINKS)) {
+            if (Files.isDirectory(fullPath)) {
+                CollectionResource c = new CollectionResource();
+                handler.handle(c);
+            } else if (Files.isRegularFile(fullPath, LinkOption.NOFOLLOW_LINKS)) {
+                putFile(handler, fullPath);
+            } else {
+                Resource r = new Resource();
+                r.exists = false;
+                handler.handle(r);
+            }
+        } else {
+            if (Files.exists(fullPath)) {
+                putFile(handler, fullPath);
+            } else {
+                try {
+                    Files.createDirectory(fullPath.getParent());
```

Comment:
Do you really have to create the parent directory?
Directories don't actually exist within S3 buckets. The entire file structure is actually just one flat single-level container of files. The illusion of directories are actually created based on naming the files names like dirA/dirB/file.
Just asking...

Reply:
I know what you mean, but it goes through java nio which needs it.